worker/upgradesteps/worker.go

```diff
@@ -310,29 +322,54 @@ func (w *upgradeSteps) runUpgradeSteps(agentConfig agent.ConfigSetter) error {
 		return errors.Trace(err)
 	}
 
-	var upgradeErr error
 	stBackend := upgrades.NewStateBackend(w.pool)
 	context := upgrades.NewContext(agentConfig, w.apiConn, stBackend)
 	logger.Infof("starting upgrade from %v to %v for %q", w.fromVersion, w.toVersion, w.tag)
 
 	targets := upgradeTargets(w.isController)
-	attempts := getUpgradeRetryStrategy()
-	for attempt := attempts.Start(); attempt.Next(); {
-		upgradeErr = PerformUpgrade(w.fromVersion, targets, context)
-		if upgradeErr == nil {
-			break
-		}
-		if agenterrors.ConnectionIsDead(logger, w.apiConn) {
-			// API connection has gone away - abort!
-			return &apiLostDuringUpgrade{upgradeErr}
-		}
-		if attempt.HasNext() {
-			w.reportUpgradeFailure(upgradeErr, true)
+
+	retryStrategy := w.retryStrategy
+	retryStrategy.IsFatalError = func(err error) bool {
+		// Abort if API connection has gone away!
+		return agenterrors.ConnectionIsDead(logger, w.apiConn)
+	}
+	retryStrategy.NotifyFunc = func(lastErr error, attempt int) {
+		if retryStrategy.Attempts != 0 && attempt != retryStrategy.Attempts {
+			w.reportUpgradeFailure(lastErr, true)
 		}
 	}
-	if upgradeErr != nil {
-		return upgradeErr
+	retryStrategy.Func = func() error {
+		err := PerformUpgrade(w.fromVersion, targets, context)
+		// w.entity.SetStatus(status.Error, fmt.Sprintf("TEST inner %v", err), nil)
+		return err
+	}
+
+	err := retry.Call(retryStrategy)
+	// w.entity.SetStatus(status.Error, fmt.Sprintf("TEST outer %v", err), nil)
+	if retry.IsAttemptsExceeded(err) || retry.IsDurationExceeded(err) {
+		err = retry.LastError(err)
+		return err
+	}
+	if err != nil {
+		return &apiLostDuringUpgrade{err}
 	}
+
+	// for attempt := attempts.Start(); attempt.Next(); {
```

Comment:
Delete.

Reply:
Oops I've been doing this a lot

At my old company, we used to comment with lightning bolt emojis when something like this was left in ⚡